File: app/backend/impl/CalendarHandler.java
package backend.impl;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import com.google.inject.ImplementedBy;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.joda.time.DateTime;
import org.joda.time.Interval;
import org.joda.time.LocalDate;
import org.joda.time.format.ISODateTimeFormat;
import play.mvc.Result;

import backend.exceptions.NotFoundException;
import backend.models.Exam;
import backend.models.ExamMachine;
import backend.models.ExamRoom;
import backend.models.Reservation;
import backend.models.User;

@ImplementedBy(CalendarHandlerImpl.class)
public interface CalendarHandler {

    Result getSlots(User user, Exam exam, Long roomId, String day, Collection<Integer> aids);
    List<ExamMachine> getEligibleMachines(ExamRoom room, Collection<Integer> access, Exam exam);
    Set<TimeSlot> handleReservations(Map<Interval, Optional<Integer>> examSlots, Collection<Reservation> reservations,
                                     Exam exam, Collection<ExamMachine> machines, User user);
    Collection<Interval> gatherSuitableSlots(ExamRoom room, LocalDate date, Integer examDuration);
    LocalDate parseSearchDate(String day, Exam exam, ExamRoom room) throws NotFoundException;
    Optional<ExamMachine> getRandomMachine(ExamRoom room, Exam exam, DateTime start, DateTime end,
                                           Collection<Integer> aids);

    Reservation createReservation(DateTime start, DateTime end, ExamMachine machine, User user,
                                  Collection<Long> sectionIds);
    LocalDate getEndSearchDate(LocalDate searchDate, LocalDate examEnd);
    int getReservationWindowSize();
    boolean isDoable(Reservation reservation, Collection<Integer> aids);

    class TimeSlot {
        private final Interval interval;
        private final String start;
        private final String end;
        private final int availableMachines;
        private final boolean ownReservation;
        private final String conflictingExam;

        public TimeSlot(Interval interval, int machineCount, String exam) {
            this.interval = interval;
            start = ISODateTimeFormat.dateTime().print(interval.getStart());
            end = ISODateTimeFormat.dateTime().print(interval.getEnd());
            availableMachines = machineCount;
            ownReservation = machineCount < 0;
            conflictingExam = exam;
        }

        Interval getInterval() {
            return interval;
        }

        public String getStart() {
            return start;
        }

        public String getEnd() {
            return end;
        }

        public int getAvailableMachines() {
            return availableMachines;
        }

        public boolean isOwnReservation() {
            return ownReservation;
        }

        public String getConflictingExam() {
            return conflictingExam;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof TimeSlot)) return false;
            TimeSlot timeSlot = (TimeSlot) o;
            return new EqualsBuilder().append(start, timeSlot.start).append(end, timeSlot.end).build();
        }

        @Override
        public int hashCode() {
            return new HashCodeBuilder().append(start).append(end).build();
        }

    }


}

File: app/services/datetime/CalendarHandler.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.datetime

import com.google.inject.ImplementedBy
import models.enrolment.{ExamEnrolment, Reservation}
import models.exam.Exam
import models.facility.{ExamMachine, ExamRoom}
import models.user.User
import play.api.libs.json.{JsValue, Json, Writes}
import play.api.mvc.Result

import java.time.format.DateTimeFormatter
import java.time.{Instant, LocalDate}
import scala.concurrent.Future

/** Error types for calendar handler operations */
sealed trait CalendarHandlerError:
  def message: String

object CalendarHandlerError:
  case class RoomNotFound(roomId: Long) extends CalendarHandlerError:
    val message = s"No room with id: ($roomId)"

@ImplementedBy(classOf[CalendarHandlerImpl])
trait CalendarHandler:
  def getSlots(
      user: User,
      exam: Exam,
      roomId: Long,
      day: String,
      aids: Seq[Long]
  ): Either[CalendarHandlerError, JsValue]

  def handleReservations(
      examSlots: Map[Interval, Option[Int]],
      reservations: Seq[Reservation],
      exam: Exam,
      machines: Seq[ExamMachine],
      user: User
  ): Set[CalendarHandler.TimeSlot]

  def gatherSuitableSlots(room: ExamRoom, date: LocalDate, examDuration: Integer): Seq[Interval]

  def parseSearchDate(day: String, exam: Exam, room: Option[ExamRoom]): Option[LocalDate]

  def getRandomMachine(
      room: ExamRoom,
      exam: Exam,
      start: Instant,
      end: Instant,
      aids: Seq[Long]
  ): Option[ExamMachine]

  def createReservation(
      start: Instant,
      end: Instant,
      machine: ExamMachine,
      user: User
  ): Reservation

  def getEndSearchDate(searchDate: LocalDate, examEnd: LocalDate): LocalDate

  def getReservationWindowSize: Int

  def isDoable(reservation: Reservation, aids: Seq[Long]): Boolean

  def handleExternalReservation(
      enrolment: ExamEnrolment,
      exam: Exam,
      node: JsValue,
      start: Instant,
      end: Instant,
      user: User,
      orgRef: String,
      roomRef: String,
      sectionIds: Seq[Long]
  ): Future[Option[Integer]]

  def postProcessSlots(
      node: JsValue,
      date: String,
      exam: Exam,
      user: User
  ): Set[CalendarHandler.TimeSlot]

  def checkEnrolment(enrolment: ExamEnrolment, user: User, sectionIds: Seq[Long]): Option[Result]

  def getEnrolment(examId: Long, user: User): ExamEnrolment

object CalendarHandler:
  case class TimeSlot(interval: Interval, machineCount: Int, exam: String):
    val start: String           = DateTimeFormatter.ISO_INSTANT.format(interval.start)
    val end: String             = DateTimeFormatter.ISO_INSTANT.format(interval.end)
    val availableMachines: Int  = machineCount
    val ownReservation: Boolean = machineCount < 0
    val conflictingExam: String = exam

  object TimeSlot:
    implicit val writes: Writes[TimeSlot] = (slot: TimeSlot) =>
      Json.obj(
        "start"             -> slot.start,
        "end"               -> slot.end,
        "availableMachines" -> slot.availableMachines,
        "ownReservation"    -> slot.ownReservation,
        "conflictingExam"   -> slot.conflictingExam
      )
